planning.py

```diff
@@ -294,6 +329,207 @@ def add_vertices(
                     if self.workspace.edge_is_in_collision(vo.coord, v.coord):
                         continue
                     vo.connect(v)
+        self.preprocessing_time = time.time() - start_time
+
+    def closest_point_not_in_collision(self, q, q_nearest):
+        """return closest point that is not in collision with workspace"""
+        while self.workspace.point_is_in_collision(q):
+            q = q + (q - q_nearest) / np.linalg.norm(q - q_nearest)
+        return q
+
+
+class RRT(RRG):
+    def __init__(self, workspace, q0):
+        super().__init__(workspace, q0)
+
+    def query(
+        self,
+        start,
+        goal,
+        n=1000,
+        min_edge_len=0.5,
+        max_edge_len=1,
+    ):
+        new_size = self.graph.n + n
+        start_time = time.time()
+        while self.graph.n < new_size:
+            # TODO maybe change min and max edge len based on how many points have been taken?
+            # add radius with minimum path
+            q = self.workspace.sample()
+
+            v_nearest, dist = self.closest_vertex(q)
+            if dist < min_edge_len:
+                continue
+            q_nearest = v_nearest.coord
+
+            # move toward q as much as possible
+            if dist > max_edge_len:
+                q = q_nearest + (q - q_nearest) * max_edge_len / dist
+            if self.workspace.point_is_in_collision(q):
+                q = self.closest_point_not_in_collision(q, q_nearest)
+            # don't add if edge is in collision
+            if self.workspace.edge_is_in_collision(q_nearest, q):
+                continue
+
+            v = self.graph.add_vertex(q)
+            v.connect(v_nearest)
+            if self.end_condition(goal):
+                self.preprocessing_time = 0
+                self.query_time = time.time() - start_time
+                break
+        self.preprocessing_time = 0
+        self.query_time = time.time() - start_time
+
+    def end_condition(self, goal, goal_dist=0.5):
+        v_nearest, dist = self.closest_vertex(goal)
+        if dist <= goal_dist and not self.workspace.edge_is_in_collision(
+            v_nearest.coord, goal
+        ):
+            v = self.graph.add_vertex(goal)
+            v.connect(v_nearest)
+            return True
+        return False
+
+
+class Double_trees(RRG):
```

Comment:
Why don't we call this class `Bidirectional RRT`? I think that is the most common name